src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.SAMSequenceRecord;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.MergingIterator;
+import htsjdk.tribble.*;
+import htsjdk.variant.variantcontext.VariantContext;
+import htsjdk.variant.variantcontext.VariantContextComparator;
+import htsjdk.variant.vcf.VCFHeader;
+import htsjdk.variant.vcf.VCFUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.SequenceDictionaryUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.io.File;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * MultiVariantDataSource aggregates multiple FeatureDataSources of variants, and enables traversals and queries
+ * over those sources through a single interface.
+ *
+ * Two basic operations are available on this data source:
+ *
+ * -Iteration over all Variants in the data sources, optionally restricted to Variants overlapping
+ *  a set of intervals if intervals are provided via {@link #setIntervalsForTraversal(List)}. Traversal
+ *  by a set of intervals requires the files to have been indexed using the bundled tool IndexFeatureFile.
+ *  The set of intervals provided MUST be non-overlapping and sorted in increasing order of start position.
+ *
+ * -Targeted queries by one interval at a time. This also requires the files to have been indexed using
+ *  the bundled tool IndexFeatureFile. Targeted queries by one interval at a time are unaffected by
+ *  any intervals for full traversal set via {@link #setIntervalsForTraversal(List)}.
+ */
+public final class MultiVariantDataSource implements GATKDataSource<VariantContext>, AutoCloseable {
+    private static final Logger logger = LogManager.getLogger(MultiVariantDataSource.class);
+
+    /**
+     * List of FeatureDataSource objects aggregated by this MultiVariantDataSource
+     */
+    private List<FeatureDataSource<VariantContext>> featureDataSources = new ArrayList<>();
+
+    /**
+     * Merged VCF header used for this (aggregate) source, derived from the individual soruces.
+     */
+    private VCFHeader mergedHeader;
+
+    /**
+     * SAMSequenceDictionary used for this (aggregate) source, derived from the individual sources.
+     */
+    private SAMSequenceDictionary mergedDictionary;
```

Comment:
Yes. This can have the side effect in some cases of emitting a sequence dictionary into the output file that was derived or partially derived from an index, and which contains 0 for sequence length (whereas in the VariantWalker-derived case no dictionary would be emitted since the input header is used). Should we not write such a sequence dictionary out ?